fixes #14664 - support IPv6 in Nic, Hosts and Hostgroups models

timogoebel opened the pull request:
This PR adds IPv6 Subnet Support to Nic, Host and Hostgroup models

Scoped search autocomplete disabled for now. It does not handle STI very well...

timogoebel committed:
fixes #14664 - support IPv6 in Nic, Hosts and Hostgroups models

lzap commented on app/models/host/managed.rb:
```diff
@@ -918,8 +923,10 @@ def recreate_config
+    if dns_ptr_record && ip4
       lookup = dns_ptr_record.dns_lookup(name_or_ip)
       return lookup.ip unless lookup.nil?
     end
```
Isnt the block 930-931 unreachable? And why not to do DNS lookup for IPv6?

Hello, thanks for your contribution :-)

lzap replied:
Oh https://github.com/theforeman/foreman/pull/3499 explains that, partially :-)

timogoebel replied:
@lzap : Good catch, that method does not make any sense at all :-) Will fix that.

lzap commented:
What exactly is disabled for scoped_search? I don't see it.

lzap commented on app/models/host/managed.rb:
```diff
@@ -742,7 +747,7 @@ def smart_proxies
   def smart_proxy_ids
     ids = []
-    [subnet, hostgroup.try(:subnet)].compact.each do |s|
+    [subnet, subnet6, hostgroup.try(:subnet), hostgroup.try(:subnet6)].compact.each do |s|
```
Nitpick: I'd perhaps do IPv4 first (host, hostgroup), then IPv6 for consistency.

lzap commented on app/models/host/managed.rb:
```diff
@@ -918,7 +923,9 @@ def recreate_config
-    return name_or_ip if name_or_ip =~ Net::Validations::IP_REGEXP
+    ip4 = Net::Validations.validate_ip(name_or_ip)
+    ip6 = Net::Validations.validate_ip6(name_or_ip)
+    return name_or_ip if ip4 || ip6
```
I believe this code isn't yet exercised. Would you mind adding a unit test?

lzap commented on test/unit/host_test.rb:
```diff
@@ -1921,66 +1923,66 @@ def teardown
       h=FactoryGirl.build(:host)
-      refute h.require_ip_validation?
+      refute h.require_ip4_validation?
     end
```
I'd still exercise `require_ip_validation` (with params) instead. Also make sure you try the IPv6 code flow which isn't exercised at all.

lzap commented on app/models/nic/base.rb:
```diff
@@ -144,16 +150,30 @@ def host_managed?
-    def require_ip_validation?
+    def require_ip_validation?(ip_field, other_ip_field, subnet)
+      raise "invalid ip field #{ip_field}" unless ip_field == :ip || ip_field == :ip6
+      raise "invalid other ip field #{other_ip_field}" unless other_ip_field == :ip || other_ip_field == :ip6
```
Please use

```
raise Foreman::Exception, '...'
```

instead which will generate an "ERF" code (http://projects.theforeman.org/projects/foreman/wiki/ErrorCodes). If this is user facing bug, then it should be also translated (http://projects.theforeman.org/projects/foreman/wiki/Translating).

lzap commented on app/models/nic/base.rb:
```diff
@@ -35,10 +35,15 @@ class Base < ActiveRecord::Base
 
-    validate :validate_host_location, :if => Proc.new { |nic| SETTINGS[:locations_enabled] && nic.subnet.present? }
-    validate :validate_host_organization, :if => Proc.new { |nic| SETTINGS[:organizations_enabled] && nic.subnet.present? }
+    validate :validate_subnets_types
```
Nitpick: Perhaps `validate_subnet_types`

lzap commented on app/models/nic/base.rb:
```diff
@@ -210,16 +230,31 @@ def sync_name
+    def validate_subnets_types
+      errors.add(:subnet, _("must be of type Subnet::Ipv4.")) if self.subnet.present? && self.subnet.type != 'Subnet::Ipv4'
+      errors.add(:subnet6, _("must be of type Subnet::Ipv6.")) if self.subnet6.present? && self.subnet6.type != 'Subnet::Ipv6'
+    end
```
Does this trigger when you enter IPv4 address into IPv6 field? I am not sure if I get when we get this error. Anyway, since this is user facing error for sure, I'd perhaps suggest something less technical (e.g. "must be an IPv6 address").

lzap commented on app/models/subnet.rb:
```diff
@@ -185,10 +185,9 @@ def unused_ip(mac = nil, excluded_ips = [])
     self.clear_association_cache
-    ips.compact.uniq
+    ips
   end
```
This does not seem to be exercised, mind adding an extra unit test for each subclass?

dLobatog replied:
Fixed

lzap commented on app/models/subnet/ipv6.rb:
```diff
@@ -23,7 +23,13 @@ def validate_ip(ip)
   def self.supported_ipam_modes
-    [:eui64, :none]
+    [:eui64, :db, :none]
+  end
```
Out of my curiousity (thank to my limited IPv6 knowledge), why IPv4 has `:dhcp` mode while IPv6 has `:eui64` mode, which I believe is also DHCP.

timogoebel replied:
The most common ways to assign an IPv6 address to a host are the following:
- Use SLAAC
- Use DHCPv6

SLAAC actually uses EUI64 to generate an IPv6 address out of the MAC Address, this totally happens on the client side so Foreman is not involved and does not have any knowledge (except for facts) of the address. DHCPv6 can use any kind of address. EUI64 is Foreman specific. Foreman generates the IPv6 address for the client. On the host it can then be set via kickstart or via Puppet. This allows Foreman to set dns aaaa records for this ip. This is good for a dual stacked approach where provisioning still happens completely via IPv4 but the hosts get an additional IPv6 address. Provisioning via IPv6 only needs some more engineering work and at least your EFI support.

lzap commented on db/migrate/20160415134454_add_ipv6_to_hosts.rb:
```diff
@@ -0,0 +1,15 @@
+    remove_column :hostgroups, :subnet6_id
+    remove_column :nics, :subnet6_id
+  end
+end
```
Is it time to introduce indices for both `:ip` and `:ip6` columns? We search during unattended installations mainly, then users do search via scoped_search. Opinion?

```
Searching 2828 files for "by_ip"

/home/lzap/work/foreman/app/controllers/unattended_controller.rb:
   77  
   78    def get_host_details
   79:     @host = find_host_by_spoof || find_host_by_token || find_host_by_ip_or_mac
   80      unless @host
   81        logger.info "#{controller_name}: unable to find a host that matches the request from #{request.env['REMOTE_ADDR']}"
   ..
   95  
   96    def find_host_by_spoof
   97:     host = Nic::Base.primary.find_by_ip(params.delete('spoof')).try(:host) if params['spoof'].present?
   98      host ||= Host.find_by_name(params.delete('hostname')) if params['hostname'].present?
   99      @spoof = host.present?
  ...
  111    end
  112  
  113:   def find_host_by_ip_or_mac
  114      # try to find host based on our client ip address
  115      ip = ip_from_request_env

/home/lzap/work/foreman/test/functional/unattended_controller_test.rb:
  174      get :built
  175      assert_response :created
  176:     host = Nic::Base.primary.find_by_ip(@ub_host.ip)
  177      assert_equal host.build,false
  178    end

/home/lzap/work/foreman/test/unit/lookup_value_test.rb:
   39      # Host.authorized(:view_hosts, Host) returns only hosts(:one)
   40      user = users(:one)
   41:     role = FactoryGirl.create(:role, :name => 'user_view_host_by_ip')
   42      FactoryGirl.create(:filter, :role => role, :permissions => [Permission.find_by_name(:view_hosts)], :search => "name = #{@host1.name}")
   43      # Todo, restore the ip test variant once our scoped-search works with host.ip again

/home/lzap/work/foreman_bootdisk/app/controllers/concerns/foreman_bootdisk/unattended_controller_ext.rb:
    3  
    4    included do
    5:     alias_method_chain :find_host_by_ip_or_mac, :param_mac
    6    end
    7  
    8:   def find_host_by_ip_or_mac_with_param_mac
    9      request.env['HTTP_X_RHN_PROVISIONING_MAC_0'] = "unknown #{params['mac']}" unless request.env.has_key?('HTTP_X_RHN_PROVISIONING_MAC_0') || params['mac'].nil?
   10:     find_host_by_ip_or_mac_without_param_mac
   11    end
   12  end

8 matches across 4 files
```

lzap commented:
The code is very good, the only comment I have overall is more unit tests. In all NIC, host and hostgroup unit tests, I'd add an extra tests for `:ip6` field added. I'd go test case by test case copying them if it also make sense for IPv6. To be honest, I don't exactly know which we are missing, spotted just few of them (e.g. primary interface detection for IPv6 deserves it's own test case).

I assume this patch is complete as you aren't touching UI or API.

timogoebel commented:
@lzap : I have added some tests, fixed some minor stuff and added the indices.

I had to disable scoped search autocomplete because it is not aware of the STI. When you search for a host via subnet it suggests both ipv4 and ipv6 subnets. I think, we can fix this - but it's not gonna be pretty so I want to do it in a separate PR.

And yes, I have UI and API prepared but @dLobatog and I agreed that we want to keep the PRs as small as possible. I find it very difficult to decide where to start and when to stop. What do you think? Does this need more love?

lzap commented on app/models/nic/base.rb:
```diff
@@ -144,16 +150,30 @@ def host_managed?
-    def require_ip_validation?
+    def require_ip_validation?(ip_field, other_ip_field, subnet)
+      raise Foreman::Exception.new(N_("invalid ip field %s") % ip_field) unless ip_field == :ip || ip_field == :ip6
+      raise Foreman::Exception.new(N_("invalid other ip field %s") % other_ip_field) unless other_ip_field == :ip || other_ip_field == :ip6
```
I apologize for such a delay. Please use this format for parameters: http://projects.theforeman.org/projects/foreman/wiki/Translating#Exceptions

timogoebel replied:
@lzap : Sorry, I don't see it... :-( What's your point exactly? Do you want to translate "ip_field"? Do we need to translate this at all? A user should never see this.

lzap replied:
> @lzap : Sorry, I don't see it... :-( What's your point exactly? Do you want to translate "ip_field"? Do we need to translate this at all? A user should never see this.

Nah, just the syntax is not right:

```
    raise Foreman::Exception.new("invalid ip field %s", ip_field)
```

As simple as that, i18n is built-in.

## 

Later,
 Lukas #lzap Zapletal

timogoebel replied:
Fixed. Thanks. 👍

lzap commented:
I apologize for the delay, I am bit late on github notifications, looks good, @dLobatog shell we merge this right away after rebase?

timogoebel commented:
Rebased.

dLobatog commented on app/models/concerns/orchestration/dhcp.rb:
```diff
@@ -174,6 +175,18 @@ def ip_belongs_to_subnet?
     # we let other validations handle that
   end
 
+  def ip6_belongs_to_subnet6?
```
This code is duplicated from ipv4, please refactor in a way there's no duplicate code. Also it's using `or` incorrectly, it should be `||` here, this is boolean "or", not control flow "or". Think about how different 

`a = false or true` is vs  `a = false || true`

dLobatog commented on app/models/nic/base.rb:
```diff
@@ -144,16 +150,30 @@ def host_managed?
+    end
+
+    def require_ip4_validation?
+      require_ip_validation?(:ip, :ip6, subnet)
```
This and the whole `require_ip_validation?` method are shadowing the `subnet` from `belongs_to` https://github.com/theforeman/foreman/pull/3498/files#diff-30afe2d935f843af2132911906f3f1c7R61 - please find another name or extract the validation from this class to make it a bit less confusing

timogoebel replied:
Renamed `subnet` to `subnet_object`.

dLobatog commented on app/models/nic/base.rb:
```diff
@@ -35,10 +35,15 @@ class Base < ActiveRecord::Base
-    validate :validate_host_organization, :if => Proc.new { |nic| SETTINGS[:organizations_enabled] && nic.subnet.present? }
+    validate :validate_subnet_types
+
+    validate :validate_host_subnet_location, :if => Proc.new { |nic| SETTINGS[:locations_enabled] && nic.host.present? && nic.subnet.present? }
```
You can define these validations in terms of `Taxonomy.enabled_taxonomies`, so that these checks are not performed on every save.

timogoebel replied:
@dLobatog: Sorry, I don't get it. Mind to elaborate a bit more? We can refactor this to make it less redundant. Do you just want to get rid of `SETTINGS[:locations_enabled]` in favor of `Taxonomy.enabled_taxonomies.include? 'locations'` ? These checks have to be performed on every save, don't they?

dLobatog commented on app/models/nic/base.rb:
```diff
@@ -210,16 +230,31 @@ def sync_name
       errors.add(:subnet, _("is not defined for host's organization.")) unless include_or_empty?(self.subnet.organizations, self.host.organization)
     end
 
+    def validate_host_subnet6_location
```
Same as above, take the whole `validate_host_subnetx_xxxxx` and define in terms of `Taxonomy.enabled_taxonomies` please

dLobatog commented on app/models/subnet/ipv4.rb:
```diff
@@ -28,6 +28,12 @@ def self.supported_ipam_modes
     [:dhcp, :db, :none]
   end
 
+  def known_ips
```
Seems like you can get rid of this and the `ipv6` equivalent by creating a class method `ip_sym` or something similar in each class and just use that in the parent class

dLobatog commented:
I left a few comments inline, please take a look @timogoebel . Overall great work here :smile: - hope we can merge it soon & finish it up!

timogoebel commented:
@lzap @dLobatog : I added two custom validators to make the validations look nicer and addressed the other comments. What do you guys think?

dLobatog commented:
Awesome job  - manual testing passed, code looks works :+1:

dLobatog commented:
Merged as 9d56081f8253c85c6ee530b4bd339390ab5c5339, thanks @timogoebel!

dLobatog commented:
Of course also major props to @lzap , :clap: thanks for reviewing most of the PR!

timogoebel commented:
@dLobatog @lzap : Thank you both for reviewing this. More PRs are incoming... :-)
